File: lib/ledger/src/engine/lots.ts
import type {
  CostMethod,
  EngineResult,
  LedgerEventInput,
  LotState,
  ProcessedEvent,
  RegistryAsset,
  ReliefRecord,
} from "../types";
import { daysBetween, exposure, proportion } from "../math";

export interface EngineOptions {
  method: CostMethod;
  /** Only events at or before this instant are applied. */
  asOf?: Date;
  /** Current multiplier per mint, used when an event does not carry its own. */
  multipliers?: Map<string, number>;
  /** Holding period in days after which a disposal counts as long term. */
  longTermDays?: number;
  resolveAsset: (mint: string) => RegistryAsset | undefined;
}

const INFLOW = new Set(["buy", "transfer_in", "wrapper_swap_in"]);
const OUTFLOW = new Set(["sell", "transfer_out", "wrapper_swap_out"]);

export function sortEvents(events: LedgerEventInput[]): LedgerEventInput[] {
  return [...events].sort((a, b) => {
    const t = a.blockTime.getTime() - b.blockTime.getTime();
    if (t !== 0) return t;
    const s = (a.slot ?? 0) - (b.slot ?? 0);
    if (s !== 0) return s;
    // Outflows before inflows inside the same transaction so wrapper swaps relieve first.
    const ao = OUTFLOW.has(a.kind) ? 0 : 1;
    const bo = OUTFLOW.has(b.kind) ? 0 : 1;
    if (ao !== bo) return ao - bo;
    return a.id.localeCompare(b.id);
  });
}

function orderLots(lots: LotState[], method: CostMethod): LotState[] {
  const open = lots.filter((l) => l.rawRemaining > 0n);
  const costPerRaw = (l: LotState): number =>
    l.costBasisUsd === null ? -1 : l.costBasisUsd / Number(l.rawQuantity);
  switch (method) {
    case "fifo":
      return open.sort((a, b) => a.openedAt.getTime() - b.openedAt.getTime() || a.id.localeCompare(b.id));
    case "lifo":
      return open.sort((a, b) => b.openedAt.getTime() - a.openedAt.getTime() || a.id.localeCompare(b.id));
    case "hifo":
      return open.sort((a, b) => {
        const diff = costPerRaw(b) - costPerRaw(a);
        if (diff !== 0) return diff;
        return a.openedAt.getTime() - b.openedAt.getTime() || a.id.localeCompare(b.id);
      });
  }
}

/**
 * Replays ledger events into lots. Pure and deterministic: the same events and
 * options always produce the same lots, reliefs and realized figures.
 */
export function runLedger(inputs: LedgerEventInput[], options: EngineOptions): EngineResult {
  const longTermDays = options.longTermDays ?? 365;
  const asOf = options.asOf;
  const events = sortEvents(inputs).filter((e) => !asOf || e.blockTime.getTime() <= asOf.getTime());
  const lotsByMint = new Map<string, LotState[]>();
  const processed: ProcessedEvent[] = [];
  const rawBalances = new Map<string, bigint>();
  const warnings: string[] = [];
  const shortfallWarned = new Set<string>();

  const multiplierFor = (e: LedgerEventInput): number =>
    e.multiplierAtEvent ?? options.multipliers?.get(e.mint) ?? 1;

  const lotsFor = (mint: string): LotState[] => {
    let lots = lotsByMint.get(mint);
    if (!lots) {
      lots = [];
      lotsByMint.set(mint, lots);
    }
    return lots;
  };

  for (const e of events) {
    const asset = options.resolveAsset(e.mint);
    const decimals = asset?.decimals ?? 0;
    const symbol = asset?.symbol ?? e.mint.slice(0, 6);
    const multiplier = multiplierFor(e);
    const exposureDelta = exposure(e.rawDelta, decimals, multiplier);

    if (e.kind !== "multiplier_change" && e.kind !== "unknown") {
      rawBalances.set(e.mint, (rawBalances.get(e.mint) ?? 0n) + e.rawDelta);
    }

    if (INFLOW.has(e.kind) && e.rawDelta > 0n) {
      const lot = openLot(e, exposureDelta, multiplier);
      lotsFor(e.mint).push(lot);
      const pricePerShare =
        lot.costBasisUsd !== null && exposureDelta > 0 ? lot.costBasisUsd / exposureDelta : null;
      processed.push({ input: e, pricePerShare, realizedUsd: null, reliefs: [] });
      continue;
    }

    if (OUTFLOW.has(e.kind) && e.rawDelta < 0n) {
      let remaining = -e.rawDelta;
      const lots = lotsFor(e.mint);
      const available = lots.reduce((acc, l) => acc + l.rawRemaining, 0n);
      if (available < remaining) {
        const shortfall = remaining - available;
        const opening: LotState = {
          id: `${e.id}:opening`,
          mint: e.mint,
          openedAt: e.blockTime,
          openSignature: null,
          openKind: "opening_balance",
          rawQuantity: shortfall,
          rawRemaining: shortfall,
          costBasisUsd: null,
          multiplierAtOpen: null,
          basisStatus: "unknown",
          basisNote: "Acquired before the indexed history. Cost basis unknown.",
          realizedUsd: 0,
          closedAt: null,
        };
        lots.push(opening);
        if (!shortfallWarned.has(e.mint)) {
          shortfallWarned.add(e.mint);
          warnings.push(
            `${symbol}: disposed more than the indexed acquisitions. An opening balance with unknown basis was created.`,
          );
        }
      }

      const proceedsTotal = disposalProceeds(e);
      const reliefs: ReliefRecord[] = [];
      const totalRaw = -e.rawDelta;
      for (const lot of orderLots(lots, options.method)) {
        if (remaining === 0n) break;
        const take = lot.rawRemaining < remaining ? lot.rawRemaining : remaining;
        const cost =
          lot.costBasisUsd === null ? null : proportion(lot.costBasisUsd, take, lot.rawQuantity);
        const proceeds = proceedsTotal === null ? 0 : proportion(proceedsTotal, take, totalRaw);
        const recognize = e.kind !== "transfer_out";
        const realized = recognize && cost !== null && proceedsTotal !== null ? proceeds - cost : null;
        lot.rawRemaining -= take;
        remaining -= take;
        if (realized !== null) lot.realizedUsd += realized;
        if (lot.rawRemaining === 0n) lot.closedAt = e.blockTime;
        reliefs.push({
          lotId: lot.id,
          rawQuantity: take,
          costBasisUsd: cost,
          proceedsUsd: proceeds,
          realizedUsd: realized,
          term: daysBetween(lot.openedAt, e.blockTime) > longTermDays ? "long" : "short",
          openedAt: lot.openedAt,
        });
      }
      const allKnown = reliefs.every((r) => r.realizedUsd !== null);
      const realizedUsd =
        e.kind === "transfer_out" || !allKnown || reliefs.length === 0
          ? null
          : reliefs.reduce((acc, r) => acc + (r.realizedUsd ?? 0), 0);
      const pricePerShare =
        proceedsTotal !== null && exposureDelta !== 0 ? proceedsTotal / Math.abs(exposureDelta) : null;
      processed.push({ input: e, pricePerShare, realizedUsd, reliefs });
      continue;
    }

    processed.push({ input: e, pricePerShare: null, realizedUsd: null, reliefs: [] });
  }

  const lots = [...lotsByMint.values()].flat();
  return { lots, events: processed, rawBalances, warnings };
}

function openLot(e: LedgerEventInput, exposureDelta: number, multiplier: number): LotState {
  let costBasisUsd: number | null = null;
  let basisStatus: LotState["basisStatus"] = "unknown";
  let basisNote: string | null = null;
  let openKind: LotState["openKind"] = "buy";

  if (e.kind === "buy") {
    if (e.grossUsd !== null) {
      costBasisUsd = e.grossUsd + (e.feeUsd ?? 0);
      basisStatus = "complete";
    } else {
      basisNote = "Purchase found but the cash leg could not be read. Cost basis unknown.";
    }
  } else if (e.kind === "transfer_in") {
    openKind = "transfer_in";
    if (e.referencePriceUsd !== null && exposureDelta > 0) {
      costBasisUsd = e.referencePriceUsd * exposureDelta;
      basisStatus = "estimated";
      basisNote = "Received by transfer. Basis estimated at the reference price when received.";
    } else {
      basisNote = "Received by transfer. No price at receipt, cost basis unknown.";
    }
  } else {
    openKind = "wrapper_swap";
    if (e.grossUsd !== null) {
      costBasisUsd = e.grossUsd;
      basisStatus = "complete";
      basisNote = "Opened by a swap between wrappers, valued at the market value of the leg given up.";
    } else {
      basisNote = "Opened by a swap between wrappers without a readable value. Cost basis unknown.";
    }
  }

  if (e.multiplierAtEvent === null && basisStatus === "complete" && multiplier !== 1) {
    basisStatus = "estimated";
    basisNote =
      "Multiplier at acquisition not reconstructed. Per share cost assumes the current multiplier.";
  }

  return {
    id: e.id,
    mint: e.mint,
    openedAt: e.blockTime,
    openSignature: e.signature,
    openKind,
    rawQuantity: e.rawDelta,
    rawRemaining: e.rawDelta,
    costBasisUsd,
    multiplierAtOpen: e.multiplierAtEvent,
    basisStatus,
    basisNote,
    realizedUsd: 0,
    closedAt: null,
  };
}

function disposalProceeds(e: LedgerEventInput): number | null {
  // Transfers out are not disposals. The lot leaves at cost and no gain is recognized.
  if (e.kind === "transfer_out") return null;
  if (e.grossUsd === null) return null;
  return e.grossUsd - (e.feeUsd ?? 0);
}

File: audit/lots.py
"""Independent decimal lot replay for Clearbook ledger events."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone
from decimal import Decimal
from typing import Any, Iterable


ZERO = Decimal("0")
CENT = Decimal("0.01")
INFLOWS = {"buy", "transfer_in", "wrapper_swap_in"}
OUTFLOWS = {"sell", "transfer_out", "wrapper_swap_out"}


def decimal(value: Any) -> Decimal | None:
    if value is None:
        return None
    return Decimal(str(value))


def instant(value: str | datetime) -> datetime:
    if isinstance(value, datetime):
        return value
    return datetime.fromisoformat(value.replace("Z", "+00:00"))


def money(value: Decimal) -> Decimal:
    return value.quantize(CENT)


def is_long_term(opened: datetime, closed: datetime) -> bool:
    year = opened.year + 1
    month = opened.month
    day = opened.day
    if month == 2 and day == 29:
        try:
            anniversary = opened.replace(year=year)
        except ValueError:
            anniversary = opened.replace(year=year, day=28)
    else:
        anniversary = opened.replace(year=year)
    sold_day = closed.astimezone(timezone.utc).date()
    return sold_day > anniversary.astimezone(timezone.utc).date()


@dataclass
class Lot:
    id: str
    mint: str
    symbol: str
    opened_at: datetime
    kind: str
    quantity: Decimal
    remaining: Decimal
    cost: Decimal | None
    multiplier_at_open: Decimal | None
    basis_status: str
    closed_at: datetime | None = None

    @property
    def remaining_cost(self) -> Decimal | None:
        if self.cost is None:
            return None
        return self.cost * self.remaining / self.quantity


@dataclass
class ClosedLot:
    disposal_id: str
    lot_id: str
    mint: str
    symbol: str
    opened_at: datetime | None
    closed_at: datetime
    quantity: Decimal
    proceeds: Decimal | None
    cost: Decimal | None
    gain: Decimal | None
    term: str


@dataclass
class Replay:
    lots: list[Lot]
    closed: list[ClosedLot]

    def open_lots(self, multipliers: dict[str, Decimal] | None = None) -> list[dict[str, Any]]:
        current = multipliers or {}
        rows = []
        for lot in self.lots:
            if lot.remaining <= ZERO:
                continue
            multiplier = current.get(lot.mint, lot.multiplier_at_open or Decimal("1"))
            rows.append(
                {
                    "id": lot.id,
                    "mint": lot.mint,
                    "symbol": lot.symbol,
                    "quantity": lot.remaining * multiplier,
                    "cost": lot.remaining_cost,
                }
            )
        return rows


def _event_quantity(event: dict[str, Any]) -> Decimal:
    if "rawDelta" in event:
        raw = decimal(event["rawDelta"]) or ZERO
        decimals = int(event.get("decimals", 0))
        return raw / (Decimal(10) ** decimals)
    return decimal(event.get("rawQuantity")) or ZERO


def _event_multiplier(event: dict[str, Any], quantity: Decimal) -> Decimal:
    explicit = decimal(event.get("multiplierAtEvent"))
    if explicit is not None:
        return explicit
    exposure = decimal(event.get("quantity"))
    if exposure is not None and quantity:
        return abs(exposure / quantity)
    return Decimal("1")


def _sort_key(event: dict[str, Any]) -> tuple[Any, ...]:
    kind = event.get("kind", "unknown")
    return (
        instant(event["blockTime"]),
        event.get("slot") or 0,
        0 if kind in OUTFLOWS else 1,
        str(event["id"]),
    )


def replay(events: Iterable[dict[str, Any]], method: str = "fifo") -> Replay:
    if method not in {"fifo", "lifo", "hifo"}:
        raise ValueError(f"Unsupported method: {method}")
    lots: list[Lot] = []
    closed: list[ClosedLot] = []

    for event in sorted(events, key=_sort_key):
        kind = event.get("kind", "unknown")
        quantity = _event_quantity(event)
        multiplier = _event_multiplier(event, quantity)
        mint = str(event["mint"])
        symbol = str(event.get("symbol") or mint[:6])
        at = instant(event["blockTime"])
        gross = decimal(event.get("grossUsd", event.get("grossAmount")))
        fee = decimal(event.get("feeUsd", event.get("fee"))) or ZERO

        if kind in INFLOWS and quantity > ZERO:
            cost = None
            status = "unknown"
            lot_kind = "buy"
            if kind == "buy" and gross is not None:
                cost = gross + fee
                status = "complete"
            elif kind == "transfer_in":
                lot_kind = "transfer_in"
                reference = decimal(event.get("referencePriceUsd"))
                if reference is not None:
                    cost = reference * quantity * multiplier
                    status = "estimated"
            elif kind == "wrapper_swap_in":
                lot_kind = "wrapper_swap"
                if gross is not None:
                    cost = gross
                    status = "complete"
            lots.append(
                Lot(
                    str(event["id"]),
                    mint,
                    symbol,
                    at,
                    lot_kind,
                    quantity,
                    quantity,
                    cost,
                    multiplier,
                    status,
                )
            )
            continue

        if kind not in OUTFLOWS or quantity >= ZERO:
            continue

        needed = -quantity
        candidates = [lot for lot in lots if lot.mint == mint and lot.remaining > ZERO]
        available = sum((lot.remaining for lot in candidates), ZERO)
        if available < needed:
            shortfall = needed - available
            opening = Lot(
                f"{event['id']}:opening",
                mint,
                symbol,
                at,
                "opening_balance",
                shortfall,
                shortfall,
                None,
                None,
                "unknown",
            )
            lots.append(opening)
            candidates.append(opening)

        if method == "fifo":
            candidates.sort(key=lambda lot: (lot.opened_at, lot.id))
        elif method == "lifo":
            candidates.sort(key=lambda lot: lot.id)
            candidates.sort(key=lambda lot: lot.opened_at, reverse=True)
        else:
            candidates.sort(
                key=lambda lot: (
                    -(lot.cost / lot.quantity if lot.cost is not None else Decimal("-1")),
                    lot.opened_at,
                    lot.id,
                )
            )

        total = needed
        proceeds_total = None if kind == "transfer_out" or gross is None else gross - fee
        disposal_rows: list[ClosedLot] = []
        for lot in candidates:
            if needed == ZERO:
                break
            take = min(lot.remaining, needed)
            cost = None if lot.cost is None else lot.cost * take / lot.quantity
            proceeds = None if proceeds_total is None else proceeds_total * take / total
            gain = None if cost is None or proceeds is None else proceeds - cost
            lot.remaining -= take
            needed -= take
            if lot.remaining == ZERO:
                lot.closed_at = at
            if kind != "transfer_out":
                disposal_rows.append(
                    ClosedLot(
                        str(event["id"]),
                        lot.id,
                        mint,
                        symbol,
                        None if lot.kind == "opening_balance" else lot.opened_at,
                        at,
                        take * multiplier,
                        None if proceeds is None else money(proceeds),
                        None if cost is None else money(cost),
                        None,
                        "unknown" if lot.kind == "opening_balance" else ("long" if is_long_term(lot.opened_at, at) else "short"),
                    )
                )

        if proceeds_total is not None and disposal_rows:
            target = money(proceeds_total)
            listed = sum((row.proceeds or ZERO for row in disposal_rows), ZERO)
            disposal_rows[-1].proceeds = (disposal_rows[-1].proceeds or ZERO) + target - listed
        for row in disposal_rows:
            if row.proceeds is not None and row.cost is not None:
                row.gain = money(row.proceeds - row.cost)
        closed.extend(disposal_rows)

    return Replay(lots, closed)


def current_multipliers(corporate_events: Iterable[dict[str, Any]]) -> dict[str, Decimal]:
    latest: dict[str, tuple[datetime, Decimal]] = {}
    for event in corporate_events:
        value = decimal(event.get("newMultiplier"))
        at_value = event.get("effectiveAt") or event.get("detectedAt")
        if value is None or not at_value:
            continue
        at = instant(at_value)
        mint = str(event["mint"])
        if mint not in latest or at > latest[mint][0]:
            latest[mint] = (at, value)
    return {mint: value for mint, (_, value) in latest.items()}